myia/vm.py

```diff
@@ -0,0 +1,193 @@
+        while True:
+            try:
+                self.advance()
+            except StopIteration:
+                return self.values[self.graph.return_]
+
+    def tail(self):
+        """Are we in a tail call?."""
+        return self.todo[-2] is self.graph.return_
+
+    def done(self):
+        """Are we done?."""
+        return self.graph.return_ in self.values
+
+    def get_value(self, node: ANFNode):
+        """
+        Get the value for a node.
+
+        This will check if we've already computed a value for it and
+        add it to the queue otherwise. If there was no value, we stop
+        the evaluation of the current node and restart the step, since
+        no useful work was done.
+        """
+        if node not in self.values:
+            self.todo.append(node)
+            raise self.Continue()
+        return self.values[node]
+
+    def do_call(self, graph: Graph, args: List[Any]):
+        """
+        Perform a call.
+
+        This will handle tail calls correctly by reusing the current VMFrame.
+        """
+        if self.tail():
+            self.reset(graph, args)
+            raise self.Return()
+        self.calling = VMFrame(graph, args, self.env)
+
+    def advance(self):
+        """
+        Take a step.
+
+        This will perform one "step" of execution.  This corresponds
+        roughly to evaluating one node of the graph.  In the case of
+        function calls, we single-step through the called function
+        even though the apply is a single node in our graph.
+        """
+        if self.done():
+            raise StopIteration()
+
+        # This hackish section is to handle single-stepping through calls.
+        if self.calling:
+            try:
+                self.calling.advance()
+            except StopIteration:
+                retval = self.calling.values[self.graph.return_]
+                self.values[self.todo[-1]] = retval
+                self.todo.pop()
+                self.calling = None
+                return
+
+        while True:
+            node = self.todo[-1]
+            # Skip nodes that have already been evaluated
+            if node in self.values:
+                self.todo.pop()
+                continue
+
+            try:
+                self.eval_node(node)
+            except self.Continue:
+                continue
+            except self.Return:
+                return
+
+            self.todo.pop()
+            return
+
+    def eval_node(self, node: ANFNode):
+        """
+        Compute the value for a node.
+
+        This may fail and ask for a missing value to be computed or
+        enter into a call.
+        """
+        if isinstance(node, Constant):
+            self.values[node] = node.value
+        elif isinstance(node, Parameter):
+            idx = self.graph.parameters.index(node)
+            self.values[node] = self.args[idx]
+        elif isinstance(node, Apply):
+            fn = self.get_value(node.inputs[0])
```

Comment:
Since no primitive is lazy or short-circuiting with respect to its arguments, you can safely do

```python
fn, *args = [self.get_value(x) for x in node.inputs]
```

Which will simplify the code below, and will make it easier to implement a generic system for primitives.

Reply:
Right I did it this way to make sure that If didn't trigger any computing on the branch not taken, but since the branches are functions, then I could do as you say.